cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1203,21 +1204,32 @@ private newtype TAccessPath =
 private class AccessPath extends TAccessPath {
   abstract string toString();
 
-  Content getHead() { this = TCons(result, _) }
+  Content getHead() {
+    this = TConsNil(result, _)
+    or
+    this = TConsCons(result, _, _)
+  }
 
   int len() {
     this = TNil(_) and result = 0
     or
-    this = TCons(_, result)
+    this = TConsNil(_, _) and result = 1
+    or
+    this = TConsCons(_, _, result)
   }
 
   DataFlowType getType() {
     this = TNil(result)
     or
-    exists(Content head | this = TCons(head, _) | result = head.getContainerType())
+    result = this.getHead().getContainerType()
   }
 
   abstract AccessPathFront getFront();
+
+  /**
+   * Holds if `this` has `head` at the front and may be followed by `tail`.
+   */
+  abstract predicate pop(Content head, AccessPath tail);
```

Comment:
Could we make the class abstract as well?

Reply:
I would replace "Holds if `this`" with "Holds if this access path"

Reply:
The class could be abstract, but I don't see the benefit. I think it would just mean more type-checking work, either for the optimizer or at run time.

Reply:
I think non-abstract classes with abstract predicates is generally considered a bad practice, which is why I brought it up.

Reply:
Done.